### PR TITLE
Filter LabelValues calls with matchers

### PR DESCRIPTION
**Pull request by jdbaldry**
**What this PR does**:
Updates the MergeQuerable.LabelValues behavior to honor filtering with matchers. The tests that were skipped in https://github.com/cortexproject/cortex/pull/4202 are now passing and one more test is added to confirm that matchers are propagated to the downstream queryables.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by bboreham** [comment]
Am I right in thinking this change applies specifically to queries that cross multiple tenants?

File: CHANGELOG.md [hunk 1]
```diff
@@ -130,6 +130,7 @@
 * [BUGFIX] Compactor: Increment `cortex_compactor_runs_failed_total` if compactor failed compact a single tenant. #4094
 * [BUGFIX] Tracing: hot fix to avoid the Jaeger tracing client to indefinitely block the Cortex process shutdown in case the HTTP connection to the tracing backend is blocked. #4134
 * [BUGFIX] Forward proper EndsAt from ruler to Alertmanager inline with Prometheus behaviour. #4017
+* [BUGFIX] MergeQueryable: support filtering LabelValues with matchers. #4277
```
Comment (bboreham) [inline CHANGELOG.md:133]: This should say it's in the context of queries crossing multiple tenants.

File: pkg/querier/tenantfederation/merge_queryable.go [hunk 2]
```diff
@@ -200,17 +209,20 @@ type stringSliceFuncJob struct {
 	warnings storage.Warnings
 }
 
-// mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
-// on per querier in parallel. It removes duplicates and sorts the result. It
-// doesn't require the output of the stringSliceFunc to be sorted, as results
+// mergeDistinctStringSliceWithLabelValues aggregates stringSlinceFunc call
```
Comment (bboreham) [inline pkg/querier/tenantfederation/merge_queryable.go:212]:
```suggestion
// mergeDistinctStringSliceWithLabelValues aggregates stringSliceFunc call
```

File: pkg/querier/tenantfederation/merge_queryable.go [hunk 3]
```diff
@@ -200,17 +209,20 @@ type stringSliceFuncJob struct {
...
+// results from queriers whose ids match the labelValues. If a nil map is
+// provided, all queriers are used. It removes duplicates and sorts the result.
+// It doesn't require the output of the stringSliceFunc to be sorted, as results
 // of LabelValues are not sorted.
-func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
-	var jobs = make([]interface{}, len(m.ids))
+func (m *mergeQuerier) mergeDistinctStringSliceWithLabelValues(f stringSliceFunc, labelValues map[string]struct{}) ([]string, storage.Warnings, error) {
```
Comment (bboreham) [inline pkg/querier/tenantfederation/merge_queryable.go:217]: "withLabelValues" seems very generic; as I understand it the code is specifically matching tenant IDs

File: pkg/querier/tenantfederation/merge_queryable.go [hunk 4]
```diff
@@ -200,17 +209,20 @@ type stringSliceFuncJob struct {
...
+	var jobs []interface{}
 
-	for pos := range m.ids {
-		jobs[pos] = &stringSliceFuncJob{
-			querier: m.queriers[pos],
-			id:      m.ids[pos],
+	for pos, id := range m.ids {
+		if _, matched := labelValues[id]; matched || labelValues == nil {
```
Comment (bboreham) [inline pkg/querier/tenantfederation/merge_queryable.go:221]: I think I would do the nil test first

**Review by bboreham — COMMENTED** [review]
Nice to see those skipped tests disappearing!

I hadn't looked in this area of the code before, and I found this PR quite difficult to understand.
I made some suggestions about naming and descriptions to make it easier for the next person.

**Comment by jdbaldry** [comment]
Thanks for the quick review @bboreham, I will address the feedback this afternoon. To answer your question, this is indeed only relevant in the context of queries crossing multiple tenants.

File: pkg/querier/tenantfederation/merge_queryable.go (hunk 4 above)
Comment (jdbaldry) [reply pkg/querier/tenantfederation/merge_queryable.go]: Changed in 0dc3ef4

File: CHANGELOG.md (hunk 1 above)
Comment (jdbaldry) [reply CHANGELOG.md]: Changed in 6445029

File: pkg/querier/tenantfederation/merge_queryable.go (hunk 3 above)
Comment (jdbaldry) [reply pkg/querier/tenantfederation/merge_queryable.go]: I agree, I have changed this to `mergeDistinctStringSliceWithTenants` in 0dc3ef4
along with some other changes that makes it clearer that we are passing around a map of matched tenant IDs

**Comment by jdbaldry** [comment]
@bboreham hopefully things should be a little clearer now. Let me know what you think

**Review by pracucci — APPROVED** [review]
The changes look correct to me. Good job! I just left a non-blocking nit.

File: pkg/querier/tenantfederation/merge_queryable_test.go [hunk 5]
```diff
@@ -611,15 +620,12 @@ func TestMergeQueryable_LabelValues(t *testing.T) {
 					},
 					labelName:           "instance",
 					expectedLabelValues: []string{},
-					skipReason:          matchersNotImplemented,
```
[the remark this answers is not in the document]
Comment (jdbaldry) [reply pkg/querier/tenantfederation/merge_queryable_test.go:614]: Great idea, removed in dfb88df

**Review by bboreham — APPROVED** [review]
lgtm!

**Pull request merged**